tests/__init__.py

```diff
@@ -76,7 +76,18 @@ def run(cwd, exe, args, env=dict(os.environ), **kwargs):
             coverage_dir,
             *cmd,
         ]
-    return subprocess.run([*cmd, *args], cwd=cwd, env=env, **kwargs)
+    if "valgrind" in os.environ.get("RUN_ANALYZER", ""):
+        cmd = ["valgrind", "--leak-check=yes", *cmd]
+    try:
+        failed = False
+        return subprocess.run([*cmd, *args], cwd=cwd, env=env, **kwargs)
+    except subprocess.CalledProcessError:
+        cmd = " ".join(cmd)
+        args = " ".join(args)
+        cmd = f"{cmd} {args}"
```

Comment:
I get that f-strings are cute.  But `pytest.fail("running command failed: {cmd} {args}".format(cmd=" ".join(cmd), args=" ".join(args))` is perhaps as or more readable?

Reply:
I’m actually quite confused about f-strings. According to the docs, they are supposed to support expressions, but in reality that is not the case?

Reply:
https://pyformat.info/ is a reasonable good reference for what you can do.  It's only some basic item and attribute access, not full python expressions.